api/layer/object.go

```diff
@@ -222,33 +223,37 @@ func (n *layer) ListObjectsV1(ctx context.Context, p *ListObjectsParamsV1) (*Lis
 		return nil, err
 	}
 
-	allObjects, err := n.listSortedAllObjects(ctx, allObjectParams{
-		Bucket:     bkt,
-		Prefix:     p.Prefix,
-		Delimiter:  p.Delimiter,
-		StartAfter: p.Marker,
-	})
-	if err != nil {
+	if cacheKey, err = createKey(ctx, bkt.CID); err != nil {
 		return nil, err
 	}
 
-	if len(allObjects) > p.MaxKeys {
-		result.IsTruncated = true
-
-		nextObject := allObjects[p.MaxKeys-1]
-		result.NextMarker = nextObject.Name
-
-		allObjects = allObjects[:p.MaxKeys]
+	if p.Marker != "" {
+		allObjects = n.cache.GetByMarker(p.Marker, cacheKey)
 	}
 
-	for _, ov := range allObjects {
-		if ov.isDir {
-			result.Prefixes = append(result.Prefixes, ov.Name)
-		} else {
-			result.Objects = append(result.Objects, ov)
+	if allObjects == nil {
+		allObjects, err = n.listSortedAllObjects(ctx, allObjectParams{
+			Bucket:     bkt,
+			Prefix:     p.Prefix,
+			Delimiter:  p.Delimiter,
+			StartAfter: p.Marker,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if p.Marker != "" {
+			allObjects = trimAfterObjectName(p.Marker, allObjects)
```

Comment:
We don't need to trim because `listSortedAllObjects` did it.

Reply:
Hm, really, removed this